lib/Twig/Environment.php

```diff
@@ -803,18 +814,28 @@ public function getExtension($name)
      */
     public function addExtension(Twig_ExtensionInterface $extension)
     {
-        $name = $extension->getName();
+        $name = get_class($extension);
+
+        $m = new ReflectionMethod($extension, 'getName');
+        $legacyName = null;
+        // we don't trigger the deprecation notice for the internal extensions
+        // as people might have extended them (even if this is deprecated and not recommended anymore)
+        if ('Twig_Extension' !== $m->getDeclaringClass()->getName() && 'Twig_Extension_' !== substr($name, 0, 15) && 'Mock_Twig_Extension' !== substr($name, 0, 19)) {
```

Comment:
@HeahDude the core references extensions using their class names. But it must still provide the existing names in extensions for BC reasons. And so we want to avoid deprecations for this case